### ISSUE TITLE
Blog: Review ML (1)

### ISSUE THREAD

**Issue by ohmass** [ohmass/Blogs#4]
Machine Learning 可用在生活中的方方面面 譬如

当你在搜索引擎搜索内容时， ML帮助搜索引擎进行rank web pages
当你在社交媒体发图片时，ML可以分析出他们是谁并帮你tag好合照内的朋友
当你在流媒体平台观看完一部自己喜欢的剧， ML可以帮你筛选出相似的“猜你喜欢”
ML还可以帮你过滤邮件内的fraud

ML无处不在，并且已经被运用在人们的日常生活中了，只是可能我们并未发觉

现在开始 我们学习一些ML的Algorithm，并了解如何让他们表现得更好

ML是 AI 的一个sub-field

#
监督式机器学习 与 非监督式机器学习 
(Supervised Learning and Unsupervised Learning)

“Field of stuudy that gives computers the ability to learn without being explicitly programmed”
 -- Arthur Samuel (1959)

Supervised Learning:
    rapid advancements
    used most in real world applications
    
    X -> Y
    input -> output     Learns from being given "Right answers"

    1) Regression: Predict a number  infinitely many possible outputs

    2) Classification: Predict small, finite, limited set of possible output categories/classes ;
                         could be numeric/non-numeric

Unsupervised Learning:
    Data only comes with inputs x, but not output labels y. 
    Algorithm has to find structure in the data.
    1) Clustering: group similar data points together.
                    cluster a bunch of data without lable, machine decided
                    eg. Google News
    2) Anomaly detection: find unusual data points.
                            eg. email fraud
    3) Dimensionality reduction: compress data using fewer numbers.

Notation: 
  Model: f w,b(x) = wx + b
    x = "input" variable feature;
    y = "output" variable/ "target" variable
    m = num of training examples
    (x,y) = single training example
    w,b: parameters / coefficients / weights 
         (the vars that u can adjust during training in order to improve the model)

Training set: Data used to train the Model

Linear Regression Model:
    
    <insert linear_reg1.png>

Cost function:

    <insert linear_reg2.png>

    <insert C1_W1_Lab03_Cost_function_Soln-Copy1.ipynb>

Gradient descent:

    We need to find the min J(w,b), the gradient descent can help us.
    通过从一个选择的起始位置进行梯度下降，我们能找到一个最低点J(w,b).

    local minima: J(w,b) 可能不是碗形的，可能会有很多个谷底。 从不同的出发点进行梯度下降，会得到不同的最低点，这个点就叫loacal minima

    <insert gradient_des1, 2.png>

    Learning Rate: a

    <insert Learning_Rate.png>

    It can reach local minimum with fixed learning rate.
        Near a local minimum,
            - Derivative becomes smaller
            - Update steps become smaller
        Can reach minimum without decreasing learning rate a.

    
    Convex function: bowl-shaped function; only have one global minimum.

    "Batch" Gradient descent:
        <insert Batch_gd.png>

    
Multiple Features:
     f w,b(x) = w1x1 + w2x2 + ... + wnxn + b =   w dot x + b
        w = [w1, w2, w3, ..., wn]
        b is a number 

        <insert multifeatures1.png>

vectorization:
    adv: 
        1. makes code shorter
        2. code running faster

        <insert vectorization1.png>

Feature scaling:
    adv:
        1. speed up gradient descent significantly.

    <insert FS.png> * 3

Feature Engineering:
    Using intuition to design new features, by transforming or combing original features.

Polynomial Regression:
    可是non-linear的

####
<a name="4"></a>
## 4 - Refresher on linear regression

In this practice lab, you will fit the linear regression parameters $(w,b)$ to your dataset.
- The model function for linear regression, which is a function that maps from `x` (city population) to `y` (your restaurant's monthly profit for that city) is represented as 
    $$f_{w,b}(x) = wx + b$$
    

- To train a linear regression model, you want to find the best $(w,b)$ parameters that fit your dataset.  

    - To compare how one choice of $(w,b)$ is better or worse than another choice, you can evaluate it with a cost function $J(w,b)$
      - $J$ is a function of $(w,b)$. That is, the value of the cost $J(w,b)$ depends on the value of $(w,b)$.
  
    - The choice of $(w,b)$ that fits your data the best is the one that has the smallest cost $J(w,b)$.

- To find the values $(w,b)$ that gets the smallest possible cost $J(w,b)$, you can use a method called **gradient descent**. 
  - With each step of gradient descent, your parameters $(w,b)$ come closer to the optimal values that will achieve the lowest cost $J(w,b)$.
  

- The trained linear regression model can then take the input feature $x$ (city population) and output a prediction $f_{w,b}(x)$ (predicted monthly profit for a restaurant in that city).

####
